agent/lm_agent/parsing/lmx.py

```diff
@@ -0,0 +1,117 @@
+RX_IN_USE = re.compile(IN_USE_LINE)
+RX_USAGE = re.compile(USAGE_LINE)
+
+
+def parse_feature_line(line: str):
+    """
+    Parse the feature line in the LM-X output.
+    Data we need:
+    - ``feature``: license name
+    """
+    parsed_feature = RX_FEATURE.match(line)
+    if parsed_feature is None:
+        return None
+    feature_data = parsed_feature.groupdict()
+    feature_data["feature"] = feature_data["feature"]
+
+    return feature_data["feature"]
+
+
+def parse_in_use_line(line: str):
+    """
+    Paser the in use line in the LM-X output.
+    Data we need:
+    - ``in_use``: quantity of licenses being use
+    - ``total``: total amount of licenses
+
+    Obs: this line doesn't include the license name.
+    The license in use is the last one parsed before this line.
+    """
+
+    parsed_in_use = RX_IN_USE.match(line)
+    if parsed_in_use is None:
+        return None
+    in_use_data = parsed_in_use.groupdict()
+
+    return {
+        "used": int(in_use_data["in_use"]),
+        "total": int(in_use_data["total"]),
+    }
+
+
+def parse_usage_line(line: str):
+    """
+    Parse the usage line in the LS-Dyna output.
+    Data we need:
+    -``user_name``: user who booked the license
+    -``lead_host``: host using the license
+    -``booked``: quantity of licenses being used
+
+    Obs: this line also doesn't incluse the license name.
+    The license in use is the last one parsed before this line.
+    """
+    parsed_usage = RX_USAGE.match(line)
+    if parsed_usage is None:
+        return None
+    usage_data = parsed_usage.groupdict()
+
+    return {
+        "user_name": usage_data["user"],
+        "lead_host": usage_data["lead_host"],
+        "booked": int(usage_data["in_use"]),
+    }
+
+
+def parse(s: str):
+    """
+    Parse the LM-X output, using regex to match the lines we need:
+    -``feature line``: info about each license
+    -``in use line``: info about licenses in use
+    -``usage line``: info about users using licenses
+
+    Since the in use and usage line don't have the name of the license in use,
+    we're saving each parsed license in a list. This way, we can find which
+    license uthe user is using by checking the last parsed license in the list.
+    """
+    parsed_data: dict = {}
+    feature_list: list = []
+
+    for line in s.splitlines():
+        parsed_feature = parse_feature_line(line)
+        parsed_in_use = parse_in_use_line(line)
+        parsed_usage = parse_usage_line(line)
+
+        if parsed_feature:
+            parsed_data[parsed_feature] = {}
+            feature_list.append(parsed_feature)
+
+        elif parsed_in_use:
+            feature = feature_list[-1]
+            parsed_data[feature] = {
+                "used": parsed_in_use["used"],
+                "total": parsed_in_use["total"],
+                "uses": [],
+            }
+        elif parsed_usage:
+            feature = feature_list[-1]
+            parsed_data[feature]["uses"].append(parsed_usage)
```

Comment:
There's no `else` case here. Can this conditional ever have each previous expression resolve to `False`?

Reply:
Yes. The license server output contains a lot of lines with information we don't need, so we just ignore them. I'll add a else here to explicitly skip those lines, is it ok?